File: cmd/tools/websearch/schemas.go
// Package websearch handles the mcp for searching the web.
// The mcp is defined in the python scripts; this script serves as the schema
// for the Tavily mcp client, as they provide their own mcp server.
package websearch

type SearchRequest struct {
	Query             string   `json:"query"`
	SearchDepth       string   `json:"search_depth,omitempty"`
	Topic             string   `json:"topic,omitempty"`
	TimeRange         string   `json:"time_range,omitempty"`
	MaxResults        int      `json:"max_results,omitempty"`
	IncludeImages     bool     `json:"include_images,omitempty"`
	IncludeAnswer     bool     `json:"include_answer,omitempty"`
	IncludeRawContent bool     `json:"include_raw_content,omitempty"`
	IncludeDomains    []string `json:"include_domains,omitempty"`
	ExcludeDomains    []string `json:"exclude_domains,omitempty"`
	Country           string   `json:"country,omitempty"`
}

type ImageResult struct {
	URL         string `json:"url"`
	Description string `json:"description"`
}

type SearchResult struct {
	Title         string        `json:"title"`
	URL           string        `json:"url"`
	Content       string        `json:"content"`
	Score         float64       `json:"score"`
	RawContent    *string       `json:"raw_content"`
	PublishedDate *string       `json:"published_date"`
	Favicon       *string       `json:"favicon"`
	Images        []ImageResult `json:"images"`
}

type SearchResponse struct {
	Query        string         `json:"query"`
	Answer       string         `json:"answer"`
	Results      []SearchResult `json:"results"`
	Images       []ImageResult  `json:"images"`
	ResponseTime float64        `json:"response_time"`
	RequestID    string         `json:"request_id"`
}

File: cmd/tools/websearch/schemas.py
from typing import Literal, Optional

from pydantic import BaseModel, Field


class SearchInputSchema(BaseModel):
    query: str = Field(description="The search query to run.")
    dest: str | None = Field(
        default=None,
        description="Optional file path to save results as JSON. If omitted, results are returned directly.",
    )
    search_depth: Literal["basic", "advanced"] | None = Field(
        default=None,
        description="'basic' is faster; 'advanced' is slower but more thorough.",
    )
    topic: Literal["general", "news", "finance"] | None = Field(
        default=None,
        description="'general' for broad searches, 'news' for recent events, 'finance' for market data.",
    )
    time_range: Literal["day", "week", "month", "year"] | None = Field(
        default=None, description="Filter results by recency."
    )
    max_results: int | None = Field(
        default=None, description="Number of results to return, typically 1-10."
    )
    include_images: bool = Field(
        default=True,
        description="If true, image URLs relevant to the query are included.",
    )
    include_answer: bool = Field(
        default=True,
        description="If true, Tavily returns a synthesised answer string above the results.",
    )
    include_domains: list[str] = Field(
        default_factory=list,
        description="Restrict results to these domains e.g. ['github.com', 'docs.python.org'].",
    )
    exclude_domains: list[str] = Field(
        default_factory=list, description="Block results from these domains."
    )
    country: str | None = Field(
        default=None,
        description="Bias results toward a specific country code e.g. 'us', 'gb'.",
    )
    timeout: float = Field(default=60.0, description="Request timeout in seconds.")


class ImageResult(BaseModel):
    url: str
    description: str


class SearchResult(BaseModel):
    title: str
    url: str
    content: str
    score: float
    raw_content: Optional[str] = None
    published_date: Optional[str] = None
    favicon: Optional[str] = None
    images: list[str] | list[ImageResult] = []


class SearchResponseSchema(BaseModel):
    results: list[SearchResult]
    query: str
    response_time: float
    answer: str = ""
    images: list[str] | list[ImageResult] = []
    request_id: str = ""
